src/lib/comms/sol-network-impl-contiki.c

```diff
@@ -0,0 +1,183 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <errno.h>
+
+#include <contiki-net.h>
+
+#include "sol-buffer.h"
+#include "sol-log.h"
+#include "sol-network.h"
+#include "sol-util.h"
+#include "sol-vector.h"
+
+static struct sol_vector links = SOL_VECTOR_INIT(struct sol_network_link);
+
+SOL_API const char *
+sol_network_link_addr_to_str(const struct sol_network_link_addr *addr,
+    struct sol_buffer *buf)
+{
+    char *p;
+    int i;
+    bool sep = false, skipping = false, treated_zeroes = false;
+
+    SOL_NULL_CHECK(addr, NULL);
+    SOL_NULL_CHECK(buf, NULL);
+
+    if (addr->family != SOL_NETWORK_FAMILY_INET6)
+        return NULL;
+
+    if (buf->capacity - buf->used < SOL_INET_ADDR_STRLEN)
+        return NULL;
+
+    p = sol_buffer_at_end(buf);
+
+    for (i = 0; i < 16; i += 2) {
+        uint16_t part;
+
+        part = ((uint16_t)addr->addr.in6[i] << 8) | addr->addr.in6[i + 1];
+        if (part && skipping) {
+            skipping = false;
+            treated_zeroes = true;
+            sep = true;
+        }
+        if (!part && !treated_zeroes) {
+            if (!skipping) {
+                skipping = true;
+                sep = true;
+            }
+        }
+        if (sep) {
+            sol_buffer_append_char(buf, ':');
+            if (skipping)
+                sep = false;
+        }
+        if (skipping)
+            continue;
+        sol_buffer_append_printf(buf, "%" PRIx16, part);
+        sep = true;
+    }
+    if (skipping)
+        sol_buffer_append_char(buf, ':');
+
+    return p;
+}
+
+SOL_API const struct sol_network_link_addr *
+sol_network_link_addr_from_str(struct sol_network_link_addr *addr, const char *buf)
+{
+    SOL_NULL_CHECK(addr, NULL);
+    SOL_NULL_CHECK(buf, NULL);
+
+    if (addr->family != SOL_NETWORK_FAMILY_INET6)
+        return NULL;
+
+    uiplib_ip6addrconv(buf, (uip_ip6addr_t *)&addr->addr.in6);
+
+    return addr;
+}
+
+static bool
+get_local_address(struct sol_network_link_addr *addr)
+{
+    uip_ds6_addr_t *dsaddr;
+
+    dsaddr = uip_ds6_get_global(-1);
+    if (!dsaddr)
+        dsaddr = uip_ds6_get_link_local(-1);
+    SOL_NULL_CHECK(dsaddr, false);
+
+    addr->family = SOL_NETWORK_FAMILY_INET6;
+    addr->port = 0;
+    memcpy(&addr->addr.in6, &dsaddr->ipaddr, sizeof(addr->addr.in6));
+
+    return true;
+}
+
+SOL_API int
+sol_network_init(void)
+{
+    struct sol_network_link *iface;
+    struct sol_network_link_addr *addr;
+
+    iface = sol_vector_append(&links);
+    SOL_NULL_CHECK(iface, -ENOMEM);
+
+    sol_vector_init(&iface->addrs, sizeof(struct sol_network_link_addr));
+
+    addr = sol_vector_append(&iface->addrs);
+    SOL_NULL_CHECK_GOTO(addr, addr_append_error);
+
+    if (!get_local_address(addr))
+        goto get_address_error;
+
+    SOL_SET_API_VERSION(iface->api_version = SOL_NETWORK_LINK_API_VERSION; )
+    iface->index = 0;
```

Comment:
On Mon, 18 Apr 2016 11:43:08 -0700, Guilherme Íscaro wrote:
> > +    struct sol_network_link *iface;
> > +    struct sol_network_link_addr *addr;
> > +
> > +    iface = sol_vector_append(&links);
> > +    SOL_NULL_CHECK(iface, -ENOMEM);
> > +
> > +    sol_vector_init(&iface->addrs, sizeof(struct sol_network_link_addr));
> > +
> > +    addr = sol_vector_append(&iface->addrs);
> > +    SOL_NULL_CHECK_GOTO(addr, addr_append_error);
> > +
> > +    if (!get_local_address(addr))
> > +        goto get_address_error;
> > +
> > +    SOL_SET_API_VERSION(iface->api_version = SOL_NETWORK_LINK_API_VERSION; )
> > +    iface->index = 0;
> 
> This is not needed. sol_vector_append() returns a `calloc()`ed memory

I'd still be explicit here.